scan模式使用filter解决ERR SELECT is not allowed in cluster mode失败

**问题描述**

scan模式使用 **filter** 解决[ERR SELECT is not allowed in cluster mode](https://github.com/alibaba/RedisShake/issues/499) 错误失败；**使用sync模式成功**。请问对scan模式是有什么特殊配置吗
> - redis-shake版本: [redis-shake-v3.1.7](https://github.com/alibaba/RedisShake/releases/tag/v3.1.7)；
> - redis-cluster版本:  4.0
> sync.toml和scan.toml模板均在内置模板的基础上只修改了source的version和address，target的type、vesion和address

swap_db.lua
```
function filter(id, is_base, group, cmd_name, keys, slots, db_id, timestamp_ms)
    return 0,0
end
```

执行命令`python3 cluster_helper.py ../redis-shake ../scan.toml ../filters/swap_db.lua`
redis-shake 的日志：
- python执行日志
```
...
'slave', 'master_id': '00e6ea38d8ae12b0e33d4a614f1c3261e7521aa2', 'last_ping_sent': '0', 'last_pong_rcvd': '1670059185000', 'epoch': '4', 'slots': [], 'migrations': [], 'connected': True}}
addresses:
1.2.3.4:6371
1.2.3.4:6372
1.2.3.4:6373
start syncing...
sleep 3 seconds to wait redis-shake start
================ 2022-12-03 17:19:47 ================
get metrics from [1.2.3.4:6371] failed: [Errno Expecting value] <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html><head>
<meta http-equiv="Content-Type" CONTENT="text/html; charset=utf-8">
<title>ERROR: The requested URL could not be retrieved</title>
<style type="text/css"><!-- 
 /*
 Stylesheet for Squid Error pages
 Adapted from design by Free CSS Templates
 http://www.freecsstemplates.org
 Released for free under a Creative Commons Attribution 2.5 License
*/

/* Page basics */
* {
        font-family: verdana, sans-serif;
}
...
<div id="content">
<p>The following error was encountered while trying to retrieve the URL: <a href="http://localhost:11008/">http://localhost:11008/</a></p>

<blockquote id="error">
<p><b>Connection to 127.0.0.1 failed.</b></p>
</blockquote>

<p id="sysmsg">The system returned: <i>(111) Connection refused</i></p>

<p>The remote host or network may be down. Please try the request again.</p>
...
```

- 各节点stderr
```
panic: ERR SELECT is not allowed in cluster mode

goroutine 38 [running]:
github.com/rs/zerolog.(*Logger).Panic.func1({0xc0000c6180, 0x0})
	github.com/rs/zerolog@v1.27.0/log.go:359 +0x2d
github.com/rs/zerolog.(*Event).msg(0xc0000b2120, {0xc0000c6180, 0x29})
	github.com/rs/zerolog@v1.27.0/event.go:156 +0x2b8
github.com/rs/zerolog.(*Event).Msg(...)
	github.com/rs/zerolog@v1.27.0/event.go:108
github.com/alibaba/RedisShake/internal/log.logFinally(0xc0000b2120, {0xc0000c6150, 0xc00029a180}, {0x0, 0xc000080ed0, 0x669d9e})
	github.com/alibaba/RedisShake/internal/log/func.go:77 +0x53
github.com/alibaba/RedisShake/internal/log.Panicf({0xc0000c6150, 0x29}, {0x0, 0x0, 0x0})
	github.com/alibaba/RedisShake/internal/log/func.go:27 +0x57
github.com/alibaba/RedisShake/internal/log.PanicError({0x81c3a0, 0xc0000be4c0})
	github.com/alibaba/RedisShake/internal/log/func.go:31 +0x33
github.com/alibaba/RedisShake/internal/client.(*Redis).DoWithStringReply(0xc000260140, {0xc000080f98, 0x2, 0x4e8668})
	github.com/alibaba/RedisShake/internal/client/redis.go:71 +0x50
github.com/alibaba/RedisShake/internal/reader.(*scanReader).scan(0xc000250300)
	github.com/alibaba/RedisShake/internal/reader/scan_reader.go:52 +0xb5
created by github.com/alibaba/RedisShake/internal/reader.(*scanReader).StartRead
	github.com/alibaba/RedisShake/internal/reader/scan_reader.go:43 +0xc7

```
- 各节点stdout
```
...
[90m2022-12-03 16:50:15[0m [1m[31mPNC[0m[0m ERR SELECT is not allowed in cluster mode
```
---

**源端 Redis**
版本：v4.0，自建
日志：
```
Background saving terminated with success
1:M 03 Dec 08:46:07.393 * Synchronization with slave xx.xx.xx.xxx:10007 succeeded
1:M 03 Dec 08:46:13.387 # Connection with slave xx.xx.xx.xxx:10007 lost
```

---

**目的端 Redis**
版本：v4.0，自建
日志：
```
1:M 03 Dec 08:23:05.874 * Background saving terminated with success
1:M 03 Dec 08:23:05.874 * Synchronization with xx.xx.xx.xxx:6376 succeeded
```

再将toml的advance功能除dir均删除后不会出现如下错误
```
get metrics from [1.2.3.4:6371] failed: [Errno Expecting value] <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html><head>
<meta http-equiv="Content-Type" CONTENT="text/html; charset=utf-8">
<title>ERROR: The requested URL could not be retrieved</title>
<style type="text/css"><!-- 
```
但是在scan模式下依然会出现`failed: HTTPConnectionPool错误
```
python3 cluster_helper.py ../redis-shake ../scan.toml ../filters/swap_db.lua
...
================ 2022-12-05 10:58:34 ================
get metrics from [1.2.3.4:6371] failed: HTTPConnectionPool(host='localhost', port=11008): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fdd85cdf6d8>: Failed to establish a new connection: [Errno 111] Connection refused',))
1.2.3.4:6372 syncing. dbId=[0], percent=[80.20]%, allowOps=[32351.20], disallowOps=[0.00], entryId=[840586], InQueueEntriesCount=[1020], unansweredBytesCount=[1070002]bytes 
1.2.3.4:6373 syncing. dbId=[0], percent=[86.14]%, allowOps=[36871.60], disallowOps=[0.00], entryId=[902972], InQueueEntriesCount=[636], unansweredBytesCount=[1108039]bytes 
================ 2022-12-05 10:58:39 ================
get metrics from [1.2.3.4:6371] failed: HTTPConnectionPool(host='localhost', port=11008): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fdd85cdf4e0>: Failed to establish a new connection: [Errno 111] Connection refused',))
get metrics from [1.2.3.4:6372] failed: HTTPConnectionPool(host='localhost', port=11009): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fdd85cdffd0>: Failed to establish a new connection: [Errno 111] Connection refused',))
get metrics from [1.2.3.4:6373] failed: HTTPConnectionPool(host='localhost', port=11010): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fdd85cf6978>: Failed to establish a new connection: [Errno 111] Connection refused',))
no redis-shake is running
Waiting for process 14877 to exit...
process 14877 exited.
Waiting for process 14878 to exit...
process 14878 exited.
Waiting for process 14884 to exit...
process 14884 exited.
```
data目录中某个节点的stdout日志如下
```
[90m2022-12-05 11:26:53[0m [32mINF[0m GOOS: linux, GOARCH: amd64
[90m2022-12-05 11:26:53[0m [32mINF[0m Ncpu: 4, GOMAXPROCS: 4
[90m2022-12-05 11:26:53[0m [32mINF[0m pid: 19408
[90m2022-12-05 11:26:53[0m [32mINF[0m pprof_port: 0
[90m2022-12-05 11:26:53[0m [32mINF[0m metrics url: http://localhost:11008
[90m2022-12-05 11:26:53[0m [32mINF[0m no password. address=[1.2.3.4:6371]
[90m2022-12-05 11:26:53[0m [32mINF[0m redisClusterWriter load cluster nodes. line=e98c09fc0f70c2249eaab2106e8ac2b6365871e2 1.2.3.4:6371@16371 myself,master - 0 1670210813000 1 connected 0-5460
[90m2022-12-05 11:26:53[0m [32mINF[0m no password. address=[1.2.3.4:6371]
[90m2022-12-05 11:26:53[0m [32mINF[0m redisWriter connected to redis successful. address=[1.2.3.4:6371]
[90m2022-12-05 11:26:53[0m [32mINF[0m redisClusterWriter load cluster nodes. line=794f0901358d287e3a641ff4d62f7f0cd9149595 1.2.3.4:6373@16373 master - 0 1670210814066 3 connected 10923-16383
[90m2022-12-05 11:26:53[0m [32mINF[0m no password. address=[1.2.3.4:6373]
[90m2022-12-05 11:26:53[0m [32mINF[0m redisWriter connected to redis successful. address=[1.2.3.4:6373]
[90m2022-12-05 11:26:53[0m [32mINF[0m redisClusterWriter load cluster nodes. line=e38c0f23365e0de5ca985e04c18ff097abb11039 1.2.3.4:6372@16372 master - 0 1670210813063 2 connected 5461-10922
[90m2022-12-05 11:26:53[0m [32mINF[0m no password. address=[1.2.3.4:6372]
[90m2022-12-05 11:26:53[0m [32mINF[0m redisWriter connected to redis successful. address=[1.2.3.4:6372]
[90m2022-12-05 11:26:53[0m [32mINF[0m redisClusterWriter connected to redis cluster successful. addresses=[1.2.3.4:6371 1.2.3.4:6373 1.2.3.4:6372]
[90m2022-12-05 11:26:53[0m [32mINF[0m no password. address=[1.2.3.4:6371]
[90m2022-12-05 11:26:53[0m [32mINF[0m no password. address=[1.2.3.4:6371]
[90m2022-12-05 11:26:53[0m [32mINF[0m scanReader connected to redis successful. address=[1.2.3.4:6371]
[90m2022-12-05 11:26:58[0m [32mINF[0m syncing. dbId=[0], percent=[19.14]%, allowOps=[40034.60], disallowOps=[0.00], entryId=[200173], InQueueEntriesCount=[12], unansweredBytesCount=[349793]bytes
[90m2022-12-05 11:27:03[0m [32mINF[0m syncing. dbId=[0], percent=[38.08]%, allowOps=[39698.20], disallowOps=[0.00], entryId=[398696], InQueueEntriesCount=[0], unansweredBytesCount=[696327]bytes
[90m2022-12-05 11:27:08[0m [32mINF[0m syncing. dbId=[0], percent=[56.82]%, allowOps=[39465.60], disallowOps=[0.00], entryId=[596036], InQueueEntriesCount=[1021], unansweredBytesCount=[1097213]bytes
[90m2022-12-05 11:27:13[0m [32mINF[0m syncing. dbId=[0], percent=[75.61]%, allowOps=[39122.00], disallowOps=[0.00], entryId=[791660], InQueueEntriesCount=[1024], unansweredBytesCount=[1111318]bytes
[90m2022-12-05 11:27:18[0m [32mINF[0m syncing. dbId=[0], percent=[93.41]%, allowOps=[37466.20], disallowOps=[0.00], entryId=[978993], InQueueEntriesCount=[1022], unansweredBytesCount=[1097213]bytes
[90m2022-12-05 11:27:19[0m [1m[31mPNC[0m[0m ERR SELECT is not allowed in cluster mode
```
并且通过redis-full-check工具检查发现存在数据丢失，总数据量100W条左右
```
[INFO 2022-12-05-11:30:21 full_check.go:250]: wait 5 seconds before start
[INFO 2022-12-05-11:30:26 full_check.go:253]: ---------------- start 3th time compare
[INFO 2022-12-05-11:30:26 full_check.go:278]: start compare db 0
[INFO 2022-12-05-11:30:27 full_check.go:203]: stat:
times:3, db:0, finished:true
KeyScan:{260 260 0}
KeyConflictAtLast|string|lack_target|{260 260 0}

[INFO 2022-12-05-11:30:27 full_check.go:328]: --------------- finished! ----------------
all finish successfully, totally 520 key(s) and 0 field(s) conflict
```

我使用节点到集群的方式，出现如下错误
```
$ ./redis-shake scan.toml ./filters/swap_db.lua
...
2022-12-05 19:47:48 INF syncing. dbId=[0], percent=[24.60]%, allowOps=[51667.00], disallowOps=[0.00], entryId=[258334], InQueueEntriesCount=[0], unansweredBytesCount=[0]bytes
2022-12-05 19:47:53 INF syncing. dbId=[0], percent=[50.39]%, allowOps=[54092.80], disallowOps=[0.00], entryId=[528819], InQueueEntriesCount=[1], unansweredBytesCount=[18469]bytes
2022-12-05 19:47:58 INF syncing. dbId=[0], percent=[76.20]%, allowOps=[54192.60], disallowOps=[0.00], entryId=[799784], InQueueEntriesCount=[15], unansweredBytesCount=[4345]bytes
2022-12-05 19:48:02 PNC ERR SELECT is not allowed in cluster mode
panic: ERR SELECT is not allowed in cluster mode

goroutine 37 [running]:
github.com/rs/zerolog.(*Logger).Panic.func1({0xc0004120c0, 0x0})
        github.com/rs/zerolog@v1.27.0/log.go:359 +0x2d
github.com/rs/zerolog.(*Event).msg(0xc0005940c0, {0xc0004120c0, 0x29})
        github.com/rs/zerolog@v1.27.0/event.go:156 +0x2b8
github.com/rs/zerolog.(*Event).Msg(...)
        github.com/rs/zerolog@v1.27.0/event.go:108
github.com/alibaba/RedisShake/internal/log.logFinally(0xc0005940c0, {0xc000412060, 0xc0001101c0}, {0x0, 0xc000078ed0, 0x669d9e})
        github.com/alibaba/RedisShake/internal/log/func.go:77 +0x53
github.com/alibaba/RedisShake/internal/log.Panicf({0xc000412060, 0x29}, {0x0, 0x0, 0x0})
        github.com/alibaba/RedisShake/internal/log/func.go:27 +0x57
github.com/alibaba/RedisShake/internal/log.PanicError({0x81c3a0, 0xc0003a21d0})
        github.com/alibaba/RedisShake/internal/log/func.go:31 +0x33
github.com/alibaba/RedisShake/internal/client.(*Redis).DoWithStringReply(0xc0002e4140, {0xc000078f98, 0x2, 0x2})
        github.com/alibaba/RedisShake/internal/client/redis.go:71 +0x50
github.com/alibaba/RedisShake/internal/reader.(*scanReader).scan(0xc0002cc2c0)
        github.com/alibaba/RedisShake/internal/reader/scan_reader.go:52 +0xb5
created by github.com/alibaba/RedisShake/internal/reader.(*scanReader).StartRead
        github.com/alibaba/RedisShake/internal/reader/scan_reader.go:43 +0xc7
```

打印启动参数后输出如下
```
addresses:
1.2.3.4:6371
1.2.3.4:6372
1.2.3.4:6373
================debug: args: ['/home/test/redis-cluster/redisshark/redis-shake', 'sync.toml', '/home/test/redis-cluster/redisshark/filters/swap_db.lua']
================debug: args: ['/home/test/redis-cluster/redisshark/redis-shake', 'sync.toml', '/root/test/redis-cluster/redisshark/filters/swap_db.lua']
================debug: args: ['/home/test/redis-cluster/redisshark/redis-shake', 'sync.toml', '/home/test/redis-cluster/redisshark/filters/swap_db.lua']
```
并且我在使用 [scan的单机到集群](https://github.com/alibaba/RedisShake/wiki/%E5%BF%AB%E9%80%9F%E5%BC%80%E5%A7%8B%EF%BC%9A%E6%95%B0%E6%8D%AE%E8%BF%81%E7%A7%BB%EF%BC%88%E4%BD%BF%E7%94%A8-scan-%E6%A8%A1%E5%BC%8F%EF%BC%89#%E5%8D%95%E6%9C%BA%E5%88%B0%E9%9B%86%E7%BE%A4a-c)方式时同样出现错误，在传输了一部分数据后报错，错误结果如下
```
$ ./redis-shake scan.toml ./filters/swap_db.lua
...
2022-12-05 19:47:48 INF syncing. dbId=[0], percent=[24.60]%, allowOps=[51667.00], disallowOps=[0.00], entryId=[258334], InQueueEntriesCount=[0], unansweredBytesCount=[0]bytes
2022-12-05 19:47:53 INF syncing. dbId=[0], percent=[50.39]%, allowOps=[54092.80], disallowOps=[0.00], entryId=[528819], InQueueEntriesCount=[1], unansweredBytesCount=[18469]bytes
2022-12-05 19:47:58 INF syncing. dbId=[0], percent=[76.20]%, allowOps=[54192.60], disallowOps=[0.00], entryId=[799784], InQueueEntriesCount=[15], unansweredBytesCount=[4345]bytes
2022-12-05 19:48:02 PNC ERR SELECT is not allowed in cluster mode
panic: ERR SELECT is not allowed in cluster mode

goroutine 37 [running]:
github.com/rs/zerolog.(*Logger).Panic.func1({0xc0004120c0, 0x0})
        github.com/rs/zerolog@v1.27.0/log.go:359 +0x2d
github.com/rs/zerolog.(*Event).msg(0xc0005940c0, {0xc0004120c0, 0x29})
        github.com/rs/zerolog@v1.27.0/event.go:156 +0x2b8
github.com/rs/zerolog.(*Event).Msg(...)
        github.com/rs/zerolog@v1.27.0/event.go:108
github.com/alibaba/RedisShake/internal/log.logFinally(0xc0005940c0, {0xc000412060, 0xc0001101c0}, {0x0, 0xc000078ed0, 0x669d9e})
        github.com/alibaba/RedisShake/internal/log/func.go:77 +0x53
github.com/alibaba/RedisShake/internal/log.Panicf({0xc000412060, 0x29}, {0x0, 0x0, 0x0})
        github.com/alibaba/RedisShake/internal/log/func.go:27 +0x57
github.com/alibaba/RedisShake/internal/log.PanicError({0x81c3a0, 0xc0003a21d0})
        github.com/alibaba/RedisShake/internal/log/func.go:31 +0x33
github.com/alibaba/RedisShake/internal/client.(*Redis).DoWithStringReply(0xc0002e4140, {0xc000078f98, 0x2, 0x2})
        github.com/alibaba/RedisShake/internal/client/redis.go:71 +0x50
github.com/alibaba/RedisShake/internal/reader.(*scanReader).scan(0xc0002cc2c0)
        github.com/alibaba/RedisShake/internal/reader/scan_reader.go:52 +0xb5
created by github.com/alibaba/RedisShake/internal/reader.(*scanReader).StartRead
        github.com/alibaba/RedisShake/internal/reader/scan_reader.go:43 +0xc7
```

scan.toml如下
```
# scan.toml

type = "scan"

[source]
version = 4.0 # redis version, such as 2.8, 4.0, 5.0, 6.0, 6.2, 7.0, ...
address = "1.2.3.4:6371"
username = "" # keep empty if not using ACL
password = "" # keep empty if no authentication is required
tls = false

[target]
type = "cluster" # "standalone" or "cluster"
version = 4.0 # redis version, such as 2.8, 4.0, 5.0, 6.0, 6.2, 7.0, ...
# When the target is a cluster, write the address of one of the nodes.
# redis-shake will obtain other nodes through the `cluster nodes` command.
address = "1.2.3.5:6371"
username = "" # keep empty if not using ACL
password = "" # keep empty if no authentication is required
tls = false

[advanced]
dir = "data"
```
swap_db.lua如下
```
--- dbid: 0 -> 1
--- dbid: 1 -> 0
--- dbid: others -> drop
function filter(id, is_base, group, cmd_name, keys, slots, db_id, timestamp_ms)
    return 0,0
end
```

> 明白了，应该过滤掉多余的 select，试试这个：
> 
> ```lua
> function filter(id, is_base, group, cmd_name, keys, slots, db_id, timestamp_ms)
>     if cmd_name == "SELECT" then
>         return 1, db_id -- disallow
>     end
>     
>     return 0, 0
> end
> ```

将`swap_db.lua`替换成过滤select的后，`./redis-shake scan.toml ./filters/swap_db.lua`报错未改变，如下
```
./redis-shake scan.toml ./filters/swap_db.lua
2022-12-06 14:59:39 INF scanReader connected to redis successful. address=[1.2.3.4:6371]
2022-12-06 14:59:44 INF syncing. dbId=[0], percent=[26.34]%, allowOps=[55500.60], disallowOps=[0.00], entryId=[277502], InQueueEntriesCount=[0], unansweredBytesCount=[16291]bytes
2022-12-06 14:59:49 INF syncing. dbId=[0], percent=[53.71]%, allowOps=[57184.40], disallowOps=[0.00], entryId=[563425], InQueueEntriesCount=[385], unansweredBytesCount=[16297]bytes
2022-12-06 14:59:54 INF syncing. dbId=[0], percent=[80.47]%, allowOps=[56294.80], disallowOps=[0.00], entryId=[844911], InQueueEntriesCount=[69], unansweredBytesCount=[6516]bytes
2022-12-06 14:59:57 PNC ERR SELECT is not allowed in cluster mode
panic: ERR SELECT is not allowed in cluster mode

goroutine 22 [running]:
github.com/rs/zerolog.(*Logger).Panic.func1({0xc000523080, 0x0})
        github.com/rs/zerolog@v1.27.0/log.go:359 +0x2d
github.com/rs/zerolog.(*Event).msg(0xc0005141e0, {0xc000523080, 0x29})
        github.com/rs/zerolog@v1.27.0/event.go:156 +0x2b8
github.com/rs/zerolog.(*Event).Msg(...)
        github.com/rs/zerolog@v1.27.0/event.go:108
github.com/alibaba/RedisShake/internal/log.logFinally(0xc0005141e0, {0xc000523020, 0xc000269640}, {0x0, 0xc000078ed0, 0x669d9e})
        github.com/alibaba/RedisShake/internal/log/func.go:77 +0x53
github.com/alibaba/RedisShake/internal/log.Panicf({0xc000523020, 0x29}, {0x0, 0x0, 0x0})
        github.com/alibaba/RedisShake/internal/log/func.go:27 +0x57
github.com/alibaba/RedisShake/internal/log.PanicError({0x81c3a0, 0xc0002f0ab0})
        github.com/alibaba/RedisShake/internal/log/func.go:31 +0x33
github.com/alibaba/RedisShake/internal/client.(*Redis).DoWithStringReply(0xc000132080, {0xc000078f98, 0x2, 0x2})
        github.com/alibaba/RedisShake/internal/client/redis.go:71 +0x50
github.com/alibaba/RedisShake/internal/reader.(*scanReader).scan(0xc0001101c0)
        github.com/alibaba/RedisShake/internal/reader/scan_reader.go:52 +0xb5
created by github.com/alibaba/RedisShake/internal/reader.(*scanReader).StartRead
        github.com/alibaba/RedisShake/internal/reader/scan_reader.go:43 +0xc7
```

--- a/README.md
+++ b/README.md
@@ -12,7 +12,7 @@ redis-shake is a tool for Redis data migration and data filtering.
 * ✅ Tested on Redis 5.0, Redis 6.0 and Redis 7.0
 * 🤗 Support custom filtering rules
 * 💪 Support large instance migration
-* 💖 Support `restore` mode and `sync` mode
+* 💖 Support `restore` mode, `sync` mode and `scan` mode
 * ☁️ Support Aliyun Redis and ElastiCache
 
 For older versions of redis-shake (support codis, twemproxy) please

--- a/internal/reader/scan_reader.go
+++ b/internal/reader/scan_reader.go
@@ -1,12 +1,19 @@
 package reader
 
 import (
+	"strconv"
+	"strings"
+
 	"github.com/alibaba/RedisShake/internal/client"
 	"github.com/alibaba/RedisShake/internal/client/proto"
 	"github.com/alibaba/RedisShake/internal/entry"
 	"github.com/alibaba/RedisShake/internal/log"
 	"github.com/alibaba/RedisShake/internal/statistics"
-	"strconv"
+)
+
+const (
+	// cluster_enabled: Indicate Redis cluster is enabled. reference from https://redis.io/commands/info/
+	clusterMode = "cluster_enabled:1"
 )
 
 type dbKey struct {
@@ -21,6 +28,7 @@ type scanReader struct {
 	// client for scan keys
 	clientScan   *client.Redis
 	innerChannel chan *dbKey
+	isCluster    bool
 
 	// client for dump keys
 	clientDump     *client.Redis
@@ -34,7 +42,15 @@ func NewScanReader(address string, username string, password string, isTls bool)
 	r.clientScan = client.NewRedisClient(address, username, password, isTls)
 	r.clientDump = client.NewRedisClient(address, username, password, isTls)
 	log.Infof("scanReader connected to redis successful. address=[%s]", address)
+
+	r.isCluster = r.IsCluster()
 	return r
+}
+
+// IsCluster is for determining whether the server is in cluster mode.
+func (r *scanReader) IsCluster() bool {
+	reply := r.clientScan.DoWithStringReply("INFO", "Cluster")
+	return strings.Contains(reply, clusterMode)
 }
 
 func (r *scanReader) StartRead() chan *entry.Entry {
@@ -46,7 +62,12 @@ func (r *scanReader) StartRead() chan *entry.Entry {
 }
 
 func (r *scanReader) scan() {
-	for dbId := 0; dbId < 16; dbId++ {
+	scanDbIdUpper := 15
+	if r.isCluster {
+		log.Infof("scanReader node are in cluster mode, only scan db 0")
+		scanDbIdUpper = 0
+	}
+	for dbId := 0; dbId <= scanDbIdUpper; dbId++ {
 		var cursor uint64 = 0
 
 		reply := r.clientScan.DoWithStringReply("SELECT", strconv.Itoa(dbId))

